tests/test_flows.py

```diff
@@ -2037,3 +2037,200 @@ def flow_with_name(foo: str = "one", bar: str = "1"):
+        assert state.type == StateType.COMPLETED
+        assert my_mock.call_args_list == [call("completed1"), call("completed2")]
+
+    def test_on_completion_hooks_work_with_async_and_sync_function(self):
+        my_mock = MagicMock()
+
+        def completed1(flow, flow_run, state):
+            my_mock("sync_completed1")
+
+        async def completed2(flow, flow_run, state):
+            my_mock("async_completed2")
+
+        @flow(on_completion=[completed1, completed2])
+        def my_flow():
+            pass
+
+        state = my_flow._run()
+        assert state.type == StateType.COMPLETED
+        assert my_mock.call_args_list == [
+            call("sync_completed1"),
+            call("async_completed2"),
+        ]
+
+
+class TestFlowHooksOnFailure:
+    def test_on_failure_hooks_run_on_failure(self):
+        my_mock = MagicMock()
+
+        def failed1(flow, flow_run, state):
+            my_mock("failed1")
+
+        def failed2(flow, flow_run, state):
+            my_mock("failed2")
+
+        @flow(on_failure=[failed1, failed2])
+        def my_flow():
+            raise Exception("oops")
+
+        state = my_flow._run()
+        assert state.type == StateType.FAILED
+        assert my_mock.call_args_list == [call("failed1"), call("failed2")]
+
+    def test_on_failure_hooks_dont_run_on_completed(self):
+        my_mock = MagicMock()
+
+        def failed1(flow, flow_run, state):
+            my_mock("failed1")
+
+        def failed2(flow, flow_run, state):
+            my_mock("failed2")
+
+        @flow(on_failure=[failed1, failed2])
+        def my_flow():
+            pass
+
+        state = my_flow._run()
+        assert state.type == StateType.COMPLETED
+        assert my_mock.call_args_list == []
+
+    def test_other_failure_hooks_run_if_a_hook_fails(self):
+        my_mock = MagicMock()
+
+        def failed1(flow, flow_run, state):
+            my_mock("failed1")
+
+        def exception_hook(flow, flow_run, state):
+            raise Exception("oops")
+
+        def failed2(flow, flow_run, state):
+            my_mock("failed2")
+
+        @flow(on_failure=[failed1, exception_hook, failed2])
+        def my_flow():
+            raise Exception("oops")
+
+        state = my_flow._run()
+        assert state.type == StateType.FAILED
+        assert my_mock.call_args_list == [call("failed1"), call("failed2")]
+
+    def test_on_failure_hooks_work_with_async_function(self):
+        my_mock = MagicMock()
+
+        async def failed1(flow, flow_run, state):
+            my_mock("failed1")
+
+        async def failed2(flow, flow_run, state):
+            my_mock("failed2")
+
+        @flow(on_failure=[failed1, failed2])
+        def my_flow():
+            raise Exception("oops")
+
+        state = my_flow._run()
+        assert state.type == StateType.FAILED
+        assert my_mock.call_args_list == [call("failed1"), call("failed2")]
+
+    def test_on_failure_hooks_work_with_async_and_sync_function(self):
```

Comment:
Generally, I'd parameterize this instead (as in the `concurrency/test_api.py` tests) so failures are clearly related to sync or async.

Reply:
Solid call - fixed